hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java

```diff
@@ -0,0 +1,179 @@
+            stoppedLatch.countDown();
+        }
+    }
+
+    private void consume() {
+        while (isConsuming()) {
+            Optional<MessageBatch> inflight = Optional.empty();
+            try {
+                logger.debug("Trying to create new batch [subscription={}].", subscription.getId());
+                MessageBatchingResult result = receiver.next(subscription);
+                inflight = of(result.getBatch());
+                inflight.ifPresent(batch -> {
+                    logger.debug("Delivering batch [subscription={}].", subscription.getId());
+                    deliver(batch, createRetryer(batch, subscription.getBatchSubscriptionPolicy()));
+                    logger.debug("Finished delivering batch [subscription={}]", subscription.getId());
+                    offsets.putAllDelivered(batch.getPartitionOffsets());
+                });
+                result.getDiscarded().forEach(m -> monitoring.markDiscarded(m, subscription, "too large"));
+            } finally {
+                logger.debug("Cleaning batch [subscription={}]", subscription.getId());
+                inflight.ifPresent(this::clean);
+            }
+        }
+    }
+
+    private Retryer<MessageSendingResult> createRetryer(MessageBatch batch, BatchSubscriptionPolicy policy) {
+        return createRetryer(batch, policy.getMessageBackoff(), policy.getMessageTtl(), policy.isRetryClientErrors());
+    }
+
+    private Retryer<MessageSendingResult> createRetryer(final MessageBatch batch, int messageBackoff, int messageTtl, boolean retryClientErrors) {
+        return RetryerBuilder.<MessageSendingResult>newBuilder()
+                .retryIfExceptionOfType(IOException.class)
+                .retryIfRuntimeException()
+                .retryIfResult(result -> isConsuming() && !result.succeeded() && shouldRetryOnClientError(retryClientErrors, result))
+                .withWaitStrategy(fixedWait(messageBackoff, MILLISECONDS))
+                .withStopStrategy(attempt -> attempt.getDelaySinceFirstAttempt() > messageTtl)
+                .withRetryListener(getRetryListener(result -> monitoring.markFailed(batch, subscription, result)))
+                .build();
+    }
+
+    private boolean shouldRetryOnClientError(boolean retryClientErrors, MessageSendingResult result) {
+        return !result.isClientError() || retryClientErrors;
+    }
+
+    private void deliver(MessageBatch batch, Retryer<MessageSendingResult> retryer) {
+        try (Timer.Context timer = hermesMetrics.subscriptionLatencyTimer(subscription).time()) {
+            MessageSendingResult result = retryer.call(() -> sender.send(batch, subscription.getEndpoint(), subscription.getBatchSubscriptionPolicy().getRequestTimeout()));
+            monitoring.markSendingResult(batch, subscription, result);
+        } catch (Exception e) {
+            logger.error(format("Batch was rejected [batch_id=%s, subscription=%s].", batch.getId(), subscription.toSubscriptionName()), e);
```

Comment:
wtf... I will try this but our logback backend was putting null in place of throwable when I looked up sources...